### PR TITLE
Vectorize arange

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38697 Vectorize arange**
* #38693 Let TensorIterator::nullary_op support check_mem_overlap option

Benchmark (gcc 8.3, Debian Buster, turbo off, Release build, Intel(R)
Xeon(R) E-2136, Parallelization using OpenMP):

```python
import timeit
for dtype in ('torch.double', 'torch.float', 'torch.uint8', 'torch.int8', 'torch.int16', 'torch.int32', 'torch.int64'):
    for n, t in [(40_000, 50000),
                (400_000, 5000)]:
        print(f'torch.arange(0, {n}, dtype={dtype}) for {t} times')
        print(timeit.timeit(f'torch.arange(0, {n}, dtype={dtype})', setup=f'import torch', number=t))
```

Before:

```
torch.arange(0, 40000, dtype=torch.double) for 50000 times
1.587841397995362
torch.arange(0, 400000, dtype=torch.double) for 5000 times
0.47885190199303906
torch.arange(0, 40000, dtype=torch.float) for 50000 times
1.5519152240012772
torch.arange(0, 400000, dtype=torch.float) for 5000 times
0.4733216500026174
torch.arange(0, 40000, dtype=torch.uint8) for 50000 times
1.426058754004771
torch.arange(0, 400000, dtype=torch.uint8) for 5000 times
0.43596178699226584
torch.arange(0, 40000, dtype=torch.int8) for 50000 times
1.4289699140063021
torch.arange(0, 400000, dtype=torch.int8) for 5000 times
0.43451592899509706
torch.arange(0, 40000, dtype=torch.int16) for 50000 times
0.5714442400058033
torch.arange(0, 400000, dtype=torch.int16) for 5000 times
0.14837959500437137
torch.arange(0, 40000, dtype=torch.int32) for 50000 times
0.5964003179979045
torch.arange(0, 400000, dtype=torch.int32) for 5000 times
0.15676555599202402
torch.arange(0, 40000, dtype=torch.int64) for 50000 times
0.8390555799996946
torch.arange(0, 400000, dtype=torch.int64) for 5000 times
0.23184613398916554
```

After:

```
torch.arange(0, 40000, dtype=torch.double) for 50000 times
0.6895066159922862
torch.arange(0, 400000, dtype=torch.double) for 5000 times
0.16820953000569716
torch.arange(0, 40000, dtype=torch.float) for 50000 times
1.3640095089940587
torch.arange(0, 400000, dtype=torch.float) for 5000 times
0.39255041000433266
torch.arange(0, 40000, dtype=torch.uint8) for 50000 times
0.3422072059911443
torch.arange(0, 400000, dtype=torch.uint8) for 5000 times
0.0605111670010956
torch.arange(0, 40000, dtype=torch.int8) for 50000 times
0.3449254590086639
torch.arange(0, 400000, dtype=torch.int8) for 5000 times
0.06115841199061833
torch.arange(0, 40000, dtype=torch.int16) for 50000 times
0.7745441729930462
torch.arange(0, 400000, dtype=torch.int16) for 5000 times
0.22106765500211623
torch.arange(0, 40000, dtype=torch.int32) for 50000 times
0.720475220005028
torch.arange(0, 400000, dtype=torch.int32) for 5000 times
0.20230313099455088
torch.arange(0, 40000, dtype=torch.int64) for 50000 times
0.8144655400101328
torch.arange(0, 400000, dtype=torch.int64) for 5000 times
0.23762561299372464
```

Differential Revision: [D22291236](https://our.internmc.facebook.com/intern/diff/D22291236)

### CODE REVIEW

**Comment by ezyang** [comment]
I notice that you have also serialized the kernel here, but the original was parallel. Can't the vectorized version also be parallelized?

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 1]
```diff
@@ -12,6 +13,30 @@ namespace {
...
+    auto step = scalar_step.to<accscalar_t>();
+    at::parallel_for(0, steps, internal::GRAIN_SIZE, [&](int64_t p_begin, int64_t p_end) {
+      int64_t idx(p_begin);
+      TensorIterator it(iter);
+      cpu_serial_kernel_vec(
+          it,
+          [start, step, steps, &idx]() -> scalar_t {
+            return start + step * (idx++);
+          },
+          [start, step, steps, &idx]() -> Vec256<scalar_t> {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:30]: OK, I understand what you are saying now: you are saying that `cpu_kernel_vec` doesn't actually provide index, which indeed does seem to be true.  Alright, point is taken.

**Comment by ezyang** [comment]
OK, I guess this follows the precedent set by linspace. I'll give @VitalyFedyunin a chance to take a look.

**Comment by xuhdev** [comment]
Rebased

**Comment by xuhdev** [comment]
@VitalyFedyunin Could you review this :)

**Comment by xuhdev** [comment]
This PR has been sitting here for more than one month... Could someone review this, please?

**Comment by VitalyFedyunin** [comment]
Sorry it required some time to confirm my suspicion: 

```
o = torch.rand(3, 3, 300000).int()
o = o.permute(2, 0, 1)
torch.arange(0, 300000*3*3, out=o)
```

Will end-up with incorrect result. For example in compare to CUDA or older code.

```
o = torch.rand(3, 3, 300000).int().cuda()
o = o.permute(2, 0, 1)
torch.arange(0, 300000*3*3, out=o)
```

**Comment by xuhdev** [comment]
This is very weird. I investigated into this, and it turns out that `is_contiguous` will always judge it as contiguous, no matter what the output is: https://github.com/pytorch/pytorch/blob/078669f6c3b0eb48cdb32be4bd4c3e7a9500e682/aten/src/ATen/native/cpu/Loops.h#L275

**Comment by xuhdev** [comment]
It turns out this isn't a bug that is specific to this PR. I've reported the issue at #41047

**Comment by ngimel** [comment]
imo this example
```
o = torch.rand(3, 3, 300000).int().cuda()
o = o.permute(2, 0, 1)
torch.arange(0, 300000*3*3, out=o)
```
where o is still 3d tensor even after the `arange` op is a demonstration of incorrect resizing behavior for `out` args, and should be changed. Issue to track resizing behavior https://github.com/pytorch/pytorch/issues/41027

**Comment by xuhdev** [comment]
I believe #20342 is the reason why the CI still fails.

**Comment by xuhdev** [comment]
@VitalyFedyunin I added back the contiguity test that I have removed in this PR, and all tests have passed now.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by xuhdev** [comment]
Can we merge this now?

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 2]
```diff
@@ -12,6 +13,30 @@ namespace {
...
+    using accscalar_t = at::acc_type<scalar_t, false>;
+    auto start = scalar_start.to<accscalar_t>();
+    auto steps = scalar_steps.to<accscalar_t>();
+    auto step = scalar_step.to<accscalar_t>();
+    at::parallel_for(0, steps, internal::GRAIN_SIZE, [&](int64_t p_begin, int64_t p_end) {
+      int64_t idx(p_begin);
+      TensorIterator it(iter);
+      cpu_serial_kernel_vec(
+          it,
+          [start, step, steps, &idx]() -> scalar_t {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:27]: Hello! Sorry it fails internal hard linters with

```
./aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:27:25: error: lambda capture 'steps' is not used [-Werror,-Wunused-lambda-capture]
```

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 3]
```diff
@@ -12,6 +13,30 @@ namespace {
...
+            return start + step * (idx++);
+          },
+          [start, step, steps, &idx]() -> Vec256<scalar_t> {
+            Vec256<scalar_t> res;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:31]: Hello! Sorry it fails internal hard linters with

```
./aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:27:25: error: lambda capture 'steps' is not used [-Werror,-Wunused-lambda-capture]
```

Comment (xuhdev) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:31]: Thanks for pointing out this! Now fixed.

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp (hunk 2 above)
Comment (xuhdev) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: @VitalyFedyunin I've pushed an update to fix this, but the internal test still failed. Could you share the new error?

**Comment by xuhdev** [comment]
Thank you @VitalyFedyunin !

**Pull request closed without merging**